File: app/src/main/java/com/softbankrobotics/qisdktutorials/ui/conversation/ConversationView.java
package com.softbankrobotics.qisdktutorials.ui.conversation;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.annotation.Nullable;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;

import com.softbankrobotics.qisdktutorials.R;

public class ConversationView extends RecyclerView {

    ConversationAdapter adapter = new ConversationAdapter();

    public ConversationView(Context context) {
        this(context, null);
    }

    public ConversationView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ConversationView(Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        setup();
    }

    public void addLine(String text, ConversationItemType type) {
        adapter.addItem(text, type);
        scrollToPosition(adapter.getItemCount() - 1);
    }

    private void setup() {
        setLayoutManager(new LinearLayoutManager(getContext()));
        setAdapter(adapter);

        Drawable drawable = getContext().getDrawable(R.drawable.empty_divider);
        if (drawable != null) {
            DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(getContext(), DividerItemDecoration.VERTICAL);
            dividerItemDecoration.setDrawable(drawable);
            this.addItemDecoration(dividerItemDecoration);
        }
    }
}

File: app/src/main/java/com/softbankrobotics/qisdktutorials/ui/conversation/ConversationView.kt
/*
 * Copyright (C) 2018 Softbank Robotics Europe
 * See COPYING for the license
 */

package com.softbankrobotics.qisdktutorials.ui.conversation

import android.content.Context
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.util.AttributeSet

import com.aldebaran.qi.sdk.`object`.conversation.ConversationStatus
import com.softbankrobotics.qisdktutorials.R

class ConversationView constructor(context: Context, attrs: AttributeSet? = null, defStyle: Int = 0) : RecyclerView(context, attrs, defStyle) {

    private val adapter = ConversationAdapter()

    init {
        setup()
    }

    constructor(context: Context, attrs: AttributeSet?): this(context, attrs, 0)

    constructor(context: Context): this(context, null)

    fun bindConversationTo(conversationStatus: ConversationStatus): ConversationBinder = ConversationBinder.binding(conversationStatus, this)

    fun addLine(text: String, type: ConversationItemType) {
        adapter.addItem(text, type)
        scrollToPosition(adapter.itemCount - 1)
    }

    private fun setup() {
        layoutManager = LinearLayoutManager(context)
        setAdapter(adapter)

        val drawable = context.getDrawable(R.drawable.empty_divider_big)
        if (drawable != null) {
            val dividerItemDecoration = DividerItemDecoration(context, DividerItemDecoration.VERTICAL)
            dividerItemDecoration.setDrawable(drawable)
            this.addItemDecoration(dividerItemDecoration)
        }
    }
}
